### PR TITLE
22295 Allow user define what is failure in terms on circuit breaker

### PR DESCRIPTION
**Pull request by qingwei91**
This PR is to address #22295 

I overloaded `withCircuitBreaker` and it's siblings with an additional argument which allow user to specify what result should increase failure count and what should not.

Default behaviour is not changed, thus should be backward compatible

### CODE REVIEW

**Comment by jrudolph** [comment]
OK TO TEST

**Comment by buaya91** [comment]
Hi, @jrudolph 

The CI have been in pending state for almost a week, do you know why is that so?

**Comment by jrudolph** [comment]
PLS BUILD

**Comment by jrudolph** [comment]
@buaya91 it seems to hang from time to time.

File: akka-actor-tests/src/test/scala/akka/pattern/CircuitBreakerSpec.scala [hunk 1]
```diff
@@ -144,6 +174,16 @@ class CircuitBreakerSpec extends AkkaSpec with BeforeAndAfter {
       breaker().currentFailureCount should ===(1)
     }
 
+    "increment failure count on even number" when {
+      "even number is considered failure" in {
+        val breaker = CircuitBreakerSpec.longCallTimeoutCb()
+        breaker().currentFailureCount should ===(0)
+        breaker().withSyncCircuitBreaker(2, CircuitBreakerSpec.evenNumberIsFailure)
```
Comment (patriknw) [inline akka-actor-tests/src/test/scala/akka/pattern/CircuitBreakerSpec.scala]: Here is a good example of where it was difficult to solve without this feature. 
Triggering failure in the CB but still return `2` from the call.

File: akka-actor/src/main/scala/akka/pattern/CircuitBreaker.scala [hunk 2]
```diff
@@ -52,8 +58,17 @@ object CircuitBreaker {
...
    */
-  def create(scheduler: Scheduler, maxFailures: Int, callTimeout: FiniteDuration, resetTimeout: FiniteDuration): CircuitBreaker =
+  def create(
+    scheduler:    Scheduler,
+    maxFailures:  Int,
+    callTimeout:  FiniteDuration,
+    resetTimeout: FiniteDuration): CircuitBreaker =
     apply(scheduler, maxFailures, callTimeout, resetTimeout)
+
+  def exceptionAsFailure[T]: Try[T] ⇒ Boolean = {
```
Comment (patriknw) [inline akka-actor/src/main/scala/akka/pattern/CircuitBreaker.scala]: this can be a `val`

Comment (patriknw) [inline akka-actor/src/main/scala/akka/pattern/CircuitBreaker.scala]: what would be the corresponding thing for `Java API`?

File: akka-actor/src/main/scala/akka/pattern/CircuitBreaker.scala [hunk 3]
```diff
@@ -125,7 +159,11 @@ class CircuitBreaker(
    */
   @inline
   private[this] def swapState(oldState: State, newState: State): Boolean =
-    Unsafe.instance.compareAndSwapObject(this, AbstractCircuitBreaker.stateOffset, oldState, newState)
+    Unsafe.instance.compareAndSwapObject(
+      this,
+      AbstractCircuitBreaker.stateOffset,
+      oldState,
+      newState)
```
Comment (patriknw) [inline akka-actor/src/main/scala/akka/pattern/CircuitBreaker.scala]: these re-formatting make it difficult to review, and I can't say that this reads better

File: akka-actor/src/main/scala/akka/pattern/CircuitBreaker.scala [hunk 4]
```diff
@@ -176,12 +252,28 @@ class CircuitBreaker(
...
+   * Java API (8) for [[#withCircuitBreaker]]
+   *
+   * @param body Call needing protected
+   * @param defineFailureFn function that define what should be consider failure and thus increase failure count
+   * @return [[java.util.concurrent.CompletionStage]] containing the call result or a
+   *   `scala.concurrent.TimeoutException` if the call timed out
+   */
+  def callWithCircuitBreakerCS[T](
+    body:            Callable[CompletionStage[T]],
+    defineFailureFn: Try[T] ⇒ Boolean): CompletionStage[T] =
```
Comment (patriknw) [inline akka-actor/src/main/scala/akka/pattern/CircuitBreaker.scala]: Java API should use `java.util.Function`, and `Try` is rather alien in Java API. Not sure what we shall use?

File: akka-actor/src/main/scala/akka/pattern/CircuitBreaker.scala [hunk 5]
```diff
@@ -195,16 +287,55 @@ class CircuitBreaker(
    */
   def withSyncCircuitBreaker[T](body: ⇒ T): T =
     Await.result(
-      withCircuitBreaker(try Future.successful(body) catch { case NonFatal(t) ⇒ Future.failed(t) }),
+      withCircuitBreaker(try Future.successful(body)
+      catch {
+        case NonFatal(t) ⇒ Future.failed(t)
+      }),
```
Comment (patriknw) [inline akka-actor/src/main/scala/akka/pattern/CircuitBreaker.scala]: DRY?
```
withSyncCircuitBreaker(body, CircuitBreaker.exceptionAsFailure)
```

**Review by patriknw — CHANGES_REQUESTED** [review]
The Java API would need some more thought.
Reverting the formatting changes would simplify reviewing.

File: akka-actor/src/main/scala/akka/pattern/CircuitBreaker.scala (hunk 3 above)
Comment (buaya91) [inline akka-actor/src/main/scala/akka/pattern/CircuitBreaker.scala]: Sorry, my bad, I will revert

File: akka-actor/src/main/scala/akka/pattern/CircuitBreaker.scala (hunk 2 above)
Comment (buaya91) [inline akka-actor/src/main/scala/akka/pattern/CircuitBreaker.scala]: Great question, do you a BiFunction would cut it?

I ain't familiar with Java, but seems CompletionStage can be handle by a Bifunction according to [docs](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/CompletionStage.html#handle-java.util.function.BiFunction-)

We can allow user to pass a BiFunction which return boolean

**Comment by buaya91** [comment]
@patriknw @jrudolph 

This changes did not pass binary compatibility, I presume because I've added new methods?
And it should be harmless, correct me if I am wrong

**Comment by jrudolph** [comment]
The failures are:

```
[info] akka-actor: found 2 potential binary incompatibilities while checking against com.typesafe.akka:akka-actor_2.11:2.4.10  (filtered 50)
[error]  * method callThrough(scala.Function0,scala.Function1)scala.concurrent.Future in trait akka.pattern.CircuitBreaker#State is present only in current version
[error]    filter with: ProblemFilters.exclude[ReversedMissingMethodProblem]("akka.pattern.CircuitBreaker#State.callThrough")
[error]  * method invoke(scala.Function0,scala.Function1)scala.concurrent.Future in trait akka.pattern.CircuitBreaker#State is present only in current version
[error]    filter with: ProblemFilters.exclude[ReversedMissingMethodProblem]("akka.pattern.CircuitBreaker#State.invoke")
```

So, as long as there are no custom implementations of `CircuitBreaker` / `CircuitBreaker.State` it may be ok.

File: akka-actor/src/main/scala/akka/pattern/CircuitBreaker.scala (hunk 2 above)
Comment (patriknw) [inline akka-actor/src/main/scala/akka/pattern/CircuitBreaker.scala]: `BiFunction` is like `scala.Function2`, i.e. 2 parameters. Question is how we should represent `Try`? As two `Optional`? `BiFunction[Optional[T], Optional[Throwable]]`?

Perhaps we can use `akka.actor.Status` to represent the `Try` in the Java API. It has `Sucess` and `Failure` that are consumable from Java.

Comment (buaya91) [inline akka-actor/src/main/scala/akka/pattern/CircuitBreaker.scala]: I think simply 

`Bifunction<T, Throwable, Boolean> will do?`

As it's java, the user should be aware to catch null if they expect null

**Comment by buaya91** [comment]
@patriknw 
After trying to implement for java api, I think your suggestion is closer to what we want, I will present the problem here and look for suggestion from anyone.

Problem: Provide java api to determine whether circuit breaker should increase failure count or not, I have 3 possible solution so far.

a) accept a `BiFunction[Optional[T], Optional[Throwable], Boolean]`
   - Pro: Relatively Typesafe and easy interop with CompletionStage
   - Con: It does not suggest the function only take either a T or a Throwable, but never both

b) accept a `akka.actor.Status`
   - Pro: This provide semantic of either `Success` or `Failure`
   - Con: `Success` takes in `Any`, thus lose some typesafety

c) Implement another `Status` like ADT which Success is Generic instead of taking Any
   - Pro: Provide all we need
   - Con: Not sure if it worth it to add yet another class for such limited use case

d) accept a `java.util.function.Function[T, Boolean]`, we will expect user to do try catch inside the function if he want to handle error case
   - Pro: Might be more familiar with java user? also provide typesafety
   - Con: Does not expose the fact that it can throws

Personally I prefer **a** and **d** , looking for feedback or comments on this, especially people that uses java more

**Comment by patriknw** [comment]
I have indeed seen d) in some other Java API and found it so ugly. I vote for **b)**

**Comment by buaya91** [comment]
@patriknw 

is losing type-safety ok?

something like this

```
def rpc():Callable[CompletionStage[String]]
def defineFailure(status: akka.actor.Status): Boolean = {
   if (status.isInstanceOf[Success])
      status.asInstanceOf[Success].value == 2
   else
      false
}

callWithCircuitBreakerCS(rpc, defineFailure)

// rpc always return a String, but the defineFailure check on Int
```

this will compile


Anyway, I will implement 5 hours later, I shall see if we have better alternatives, if not I'll go with @patriknw 

Cheers

**Comment by patriknw** [comment]
@buaya91 yeah, perhaps not so nice, and we can use create a small class instead like:

```scala
class CallResult[T](success: T, failure: Throwable) {
  def isSuccess: Boolean = success != null
  def getSuccess: T = 
    if (isSuccess) success
    else throw new NoSuchElementException("No value present")
  
  // TODO similar for the failure
}
```

**Comment by buaya91** [comment]
@patriknw After some more thought, I think introducing another class is not the way, we will effectively implement yet another `Try`, albeit less powerful, so I went with option a)

Hope this will work.

**Comment by buaya91** [comment]
@jrudolph could you restart the pipeline again?
@patriknw any feedback on this?

cheers

File: akka-actor/src/main/scala/akka/pattern/CircuitBreaker.scala [hunk 6]
```diff
@@ -54,6 +57,29 @@ object CircuitBreaker {
    */
   def create(scheduler: Scheduler, maxFailures: Int, callTimeout: FiniteDuration, resetTimeout: FiniteDuration): CircuitBreaker =
     apply(scheduler, maxFailures, callTimeout, resetTimeout)
+
+  val exceptionAsFailure: Try[_] ⇒ Boolean = {
+    case _: Success[_] ⇒ false
+    case _             ⇒ true
+  }
+
+  def exceptionAsFailureJava[T]: BiFunction[Optional[T], Optional[Throwable], java.lang.Boolean] =
```
Comment (patriknw) [inline akka-actor/src/main/scala/akka/pattern/CircuitBreaker.scala]: that naming is rather ugly, is this needed as public API?

File: akka-actor/src/main/scala/akka/pattern/CircuitBreaker.scala [hunk 7]
```diff
@@ -54,6 +57,29 @@ object CircuitBreaker {
    */
   def create(scheduler: Scheduler, maxFailures: Int, callTimeout: FiniteDuration, resetTimeout: FiniteDuration): CircuitBreaker =
     apply(scheduler, maxFailures, callTimeout, resetTimeout)
+
+  val exceptionAsFailure: Try[_] ⇒ Boolean = {
```
Comment (patriknw) [inline akka-actor/src/main/scala/akka/pattern/CircuitBreaker.scala]: is this needed as public API?

**Review by patriknw — CHANGES_REQUESTED** [review]
this looks good, with one small naming question

**Comment by patriknw** [comment]
PLS BUILD

File: akka-actor/src/main/scala/akka/pattern/CircuitBreaker.scala (hunk 7 above)
Comment (buaya91) [inline akka-actor/src/main/scala/akka/pattern/CircuitBreaker.scala]: You're right, I will change it to private

File: akka-actor/src/main/scala/akka/pattern/CircuitBreaker.scala (hunk 6 above)
Comment (buaya91) [inline akka-actor/src/main/scala/akka/pattern/CircuitBreaker.scala]: You're right, I will change it to private

File: akka-actor/src/main/scala/akka/pattern/CircuitBreaker.scala (hunk 7 above)
Comment (buaya91) [inline akka-actor/src/main/scala/akka/pattern/CircuitBreaker.scala]: By the way, any suggestions of the name?

Comment (patriknw) [inline akka-actor/src/main/scala/akka/pattern/CircuitBreaker.scala]: If it's private I think the name is fine

**Comment by patriknw** [comment]
@buaya91 binary compatibility check failed:
```
[error]  * method callThrough(scala.Function0,scala.Function1)scala.concurrent.Future in trait akka.pattern.CircuitBreaker#State is present only in current version
[error]    filter with: ProblemFilters.exclude[ReversedMissingMethodProblem]("akka.pattern.CircuitBreaker#State.callThrough")
[error]  * method invoke(scala.Function0,scala.Function1)scala.concurrent.Future in trait akka.pattern.CircuitBreaker#State is present only in current version
[error]    filter with: ProblemFilters.exclude[ReversedMissingMethodProblem]("akka.pattern.CircuitBreaker#State.invoke")
```

That's all internal, so it's safe to add the filters. Please add the following to project/MiMa.scala in the section `bcIssuesBetween24and25`:

```
ProblemFilters.exclude[ReversedMissingMethodProblem]("akka.pattern.CircuitBreaker#State.callThrough"),
ProblemFilters.exclude[ReversedMissingMethodProblem]("akka.pattern.CircuitBreaker#State.invoke"),
```

Somewhere here: https://github.com/akka/akka/blob/master/project/MiMa.scala#L105

**Comment by patriknw** [comment]
I rebased this and opened https://github.com/akka/akka/pull/22586

**Pull request closed without merging**